src/v/wasm/wasi.cc

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "wasi.h"
+
+#include "ffi.h"
+#include "seastarx.h"
+#include "vassert.h"
+#include "vlog.h"
+#include "wasm/logger.h"
+
+#include <seastar/core/byteorder.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/print.hh>
+#include <seastar/util/later.hh>
+#include <seastar/util/log.hh>
+
+#include <fmt/chrono.h>
+#include <fmt/format.h>
+#include <fmt/ranges.h>
+
+#include <chrono>
+#include <exception>
+#include <numeric>
+#include <ratio>
+#include <sstream>
+#include <string_view>
+#include <unistd.h>
+#include <utility>
+
+namespace wasm::wasi {
+
+log_writer::log_writer(ss::sstring name, bool is_guest_stdout, ss::logger* l)
+  : _is_guest_stdout(is_guest_stdout)
+  , _name(std::move(name))
+  , _logger(l) {}
+
+log_writer log_writer::make_for_stderr(ss::sstring name, ss::logger* l) {
+    return log_writer(std::move(name), false, l);
+}
+log_writer log_writer::make_for_stdout(ss::sstring name, ss::logger* l) {
+    return log_writer(std::move(name), true, l);
+}
+
+uint32_t log_writer::write(std::string_view buf) {
+    uint32_t amt = 0;
+    while (!buf.empty()) {
+        size_t pos = buf.find('\n');
+        if (pos == std::string_view::npos) {
+            _buffer.push_back(buf);
+            return amt;
+        }
+        _buffer.push_back(buf.substr(0, pos));
+        // Add one for the newline we stripped off and flush adds
+        amt += flush() + 1;
+        buf = buf.substr(pos + 1);
+    }
+    return amt;
+}
+
+uint32_t log_writer::flush() {
+    if (_buffer.empty()) {
+        return 0;
+    }
+    auto buf = std::exchange(_buffer, {});
+    uint32_t amt = 0;
+    for (auto b : buf) {
+        amt += b.size();
+    }
+    auto level = _is_guest_stdout ? ss::log_level::info : ss::log_level::warn;
+    _logger->log(level, "{} - {}", _name, fmt::join(buf, ""));
+    return amt;
+}
+
+preview1_module::preview1_module(
+  std::vector<ss::sstring> args, const environ_map_t& environ, ss::logger* l)
+  : _args(std::move(args))
+  , _stdout_log_writer(log_writer::make_for_stdout(_args.front(), l))
+  , _stderr_log_writer(log_writer::make_for_stderr(_args.front(), l)) {
+    _environ.reserve(environ.size());
+    for (const auto& [k, v] : environ) {
+        vassert(
+          k.find("=") == ss::sstring::npos, "invalid environment key: {}", k);
```

Comment:
input checked somewhere else too?

Reply:
>it's probably better to just throw in this case?

i'm not sure what kind of error handling makes sense e.g. error code vs exception, but avoiding the hard stop is probably a good idea.

Reply:
Yeah agreed, added an exception, which means that the `start` method of the wasm engine will fail.